src/github.com/matrix-org/dendrite/appservice/routing/user.go

```diff
@@ -0,0 +1,64 @@
+package routing
+
+import (
+	"encoding/json"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/util"
+	"io/ioutil"
+	"net/http"
+)
+
+// URIToUIDResponse represents response to an AppService URI to User Id
+// (/_matrix/app/r0/user?uri={url_encoded_uri}) request
+type URIToUIDResponse struct {
+	UserID string `json:"user_id"`
+}
+
+// URIToUID implements `/_matrix/app/r0/user?uri={url_encoded_uri}`, which
+// enables users to contact App Service users directly by taking an encoded
+// URI and turning it into a Matrix ID on the homeserver.
+// https://matrix.org/docs/spec/application_service/unstable.html#user-ids
+func URIToUID(req *http.Request, cfg config.Dendrite) util.JSONResponse {
+	uri := req.URL.Query().Get("uri")
+	if uri == "" {
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: nil,
+		}
+	}
+	for _, appservice := range cfg.Derived.ApplicationServices {
+		// Check all the fields associated with each application service
+		if !appservice.IsInterestedInUserID(uri) {
+			continue
+		}
+		// call the application service
+		reqURL := "http://" + appservice.URL + "/_matrix/app/unstable/thirdparty/user/" +
```

Comment:
It's also recommended to change `"/_matrix/app/unstable/thirdparty/user/"` to a const that's set at the top of the file.